glib/src/object.rs

```diff
@@ -3267,6 +3294,87 @@ wrapper! {
     }
 }
 
+struct WeakRefNotifyData {
+    callback: Box<dyn Fn()>,
+    should_drop: AtomicBool,
+}
+
+impl WeakRefNotifyData {
+    pub fn new<F: Fn() + 'static>(f: F) -> ManuallyDrop<Pin<Box<Self>>> {
+        ManuallyDrop::new(Box::pin(WeakRefNotifyData {
+            callback: Box::new(f),
+            should_drop: AtomicBool::new(false),
+        }))
+    }
+
+    // rustdoc-stripper-ignore-next
+    /// Checks whether the data should be dropped and marks it to be dropped
+    pub fn should_drop(&self) -> bool {
+        self.should_drop.swap(true, Ordering::SeqCst)
+    }
+}
+
+// rustdoc-stripper-ignore-next
+/// A handle to disconnect a weak ref notify closure.
+pub struct WeakRefNotify {
+    object: WeakRef<Object>,
+    data: ManuallyDrop<Pin<Box<WeakRefNotifyData>>>,
+    func: gobject_ffi::GWeakNotify,
+}
+
+impl WeakRefNotify {
+    pub(crate) fn new<T: ObjectType, F: Fn() + 'static>(obj: &T, f: F) -> WeakRefNotify {
+        unsafe extern "C" fn notify_func(data: ffi::gpointer, _obj: *mut gobject_ffi::GObject) {
+            let data = Box::from_raw(data as *mut WeakRefNotifyData);
+
+            (data.callback)();
+
+            // Don't drop the data when the `WeakRefNotify` handle wasn't dropped yet
+            if !data.should_drop() {
+                Box::leak(data);
```

Comment:
If your `callback` panics then we'll have a double free :) Maybe use just a reference outside this `if` here and call `from_raw` only inside the `if`?